### ISSUE TITLE
WinRM - SQL Server DB Deployment "Additional Arguments" stopped working after last update

### ISSUE THREAD

**Issue by roqetman** [Microsoft/vsts-rm-extensions#174]
WinRM - SQL Server DB Deployment "Additional Arguments" stopped working after last update (1/5/17)
I was using "Additional Arguments" to send parameters to a script. Everything worked fine last Wednesday, but when I retested the task this morning, it was giving errors that would only occur if the Additional Parameters aren't being fed to Invoke-Sqlcmd (if I run a remote powershell script using Invoke-Sqlcmd I'm able to run my complete script (with parameters) on the server).
Please could you check if those cert changes could have inadvertently broken the sending of additional arguments?
Thanks.

**Comment by Ajay-MS** [comment]
@roqetman 
Can you please share additional arguments in same format that you are passing to task and debugs logs ( Add variable System.Debug to RD with value true)

**Comment by roqetman** [comment]
Sure, it's a script to restore an instance (I have other scripts that need parameters to be passed to them too, and all now fail).
Here are the additional parameters I had set:
-querytimeout 0 -Variable newname="MyDev", backupfile="MyDev_backup.bak", olddataname="MyDev", oldlogname="MyDev_log"

And here's the content of the script:
RESTORE DATABASE [$(newname)]   
FROM URL = 'https://xxxxx.blob.core.windows.net/backups/$(backupfile)'   
WITH CREDENTIAL = 'mycreds',
BLOCKSIZE = 512,  
MOVE N'$(olddataname)' TO N'G:\MSSQL\Data\$(newname).mdf',  NOUNLOAD,  REPLACE,  STATS = 10,
MOVE N'$(oldlogname)' TO N'F:\MSSQL\Data\$(newname).log',  NOUNLOAD,  REPLACE,  STATS = 10
GO

**Comment by Ajay-MS** [comment]
@roqetman 

Thanks for providing required info. I am able to recreate issue. Cause of this issue as recently splatting has been introduce to prevent unnecessary  expansion of arguments. In splatting argument "Variable" need to be handle in special way. I will fix it soon. I apologize for issue that you are facing.
If you have OnPrem agent I can help you to unblock you for now.

**Comment by roqetman** [comment]
Thanks for looking into this - I'd rather not have to switch all my db pipelines to have to use powershell :-)
Do you have advice as to how I can send variables to the script?
Not sure what you mean by an OnPrem agent - the VM with sql installed is in azure too.
Thanks.

**Comment by Ajay-MS** [comment]
Are you using Hosted Agent?
If yes, then I have to fix from my side only. It will take approx 2-3 days. Otherwise if agent is local  can have a small fix as workaround till fix comes via task. I will send PR soon on this.

**Comment by Ajay-MS** [comment]
@roqetman 

I have raised a PR so that you can refer it if you are using local agent and make changes in task. I have to see if this issue is being true for other arguments so this PR may take some time to merge.

PR - https://github.com/Microsoft/vsts-rm-extensions/pull/175

**Comment by roqetman** [comment]
I'm deploying to an Azure VM with WinRM installed. Is there some work-around for now, or do I just need to wait to see if the component is updated?

**Comment by genesisjschubert** [comment]
@Ajay-MS  I believe this is also the root cause for an issue I encountered starting on 1/5 with v1.3.2 of SqlDacPacDeploy.  The '$env:systemdrive' environment variable started causing my deployment to fail. Replacing it with 'C:' fixed the issue.
I specifically used that variable in the **DACPAC File** and **Publish Profile** properties of the deployment step.
If I leave the env. variable in the Publish Profile property, I get this error during deployment:

> Deployment failed on machine StagingWeb with following message : System.Exception: *** Error parsing connection string: The given path's format is not supported..

If I replace it in the Publish Profile with **C:**, but leave it in the DACPAC File property, I get this error:

> Deployment failed on machine StagingWeb with following message : System.Exception: *** Could not load package from '$env:systemdrive\...\Pulse.System.dacpac'. The given path's format is not supported.

This just tells me the order in which it is parsing through the parameters, but I relay that all for your info and anyone else that may be encountering them.

We have a local agent, but my workaround is sufficient for now. I'll be interested to know when the local agent extension will be updated.

**Comment by Ajay-MS** [comment]
@roqetman @genesisjschubert 

We have fixed this issue. Can you please verify if you are still facing this issue or its being resolved.

**Comment by genesisjschubert** [comment]
@Ajay-MS I still get the same issue as detailed in my previous message.  Is there anything I need to do to force the build agent to update to a newer/fixed version?

**Comment by roqetman** [comment]
The new version works for me (allows "additional parameters" to be passed to the script). Thanks!

**Comment by Ajay-MS** [comment]
@roqetman 
Thanks for updating that additional arguments are working fine for you.

@genesisjschubert 
No your task has been updated but issue is because of following behavior 
$env:Systemdrive variable is not get expanded at time of reading input. It get expanded at time on WinRM. From last update expansion of variables during WinRm is prevented.
If you will defined a variable in RD or in environment it will work fine for you. I will discuss with if expansion of $env variables for remote machine should be allowed and will update you.

**Comment by genesisjschubert** [comment]
@Ajay-MS Thanks for the update. I would say it is important to be able to use $env variables on remote machines, especially for when there are multiple release agents. Without $env expansion, all agents would be requireed to have the same standard system drive letter, paths, etc. Should any of the agents have differing system drive letters, there's no way to account for that in the configuration and the release would fail on those non-standard setups.  Similar failures could occur if other $env variables were used elsewhere in the configuration. Granted, a non-standard system drive letter is rare in my experience, but having that restriction does limit the flexibility and portable nature when used in conjunction with multiple agents.
We'll continue to use our workaround as we currently only have a single agent.  We do intend to expand that in the coming months and I'll pass along the fact that we must stick with standard system drive letters and paths as we build those out.

**Comment by Ajay-MS** [comment]
@genesisjschubert 
As per discussion will taking this into consideration. We will fix this issue corresponding to expansion of environment variable soon.

**Comment by Ajay-MS** [comment]
@genesisjschubert 
Closing this issue as it is being titled for additional argument. I am opening new issue corresponding to environment variable expansion. Please refer new issue for tracking
[Environment variable expansion issue](https://github.com/Microsoft/vsts-rm-extensions/issues/196)
